mesonbuild/interpreterbase/decorators.py

```diff
@@ -782,6 +782,64 @@ def log_usage_warning(self, tv: str, location: T.Optional['mparser.BaseNode']) -
         mlog.deprecation(*args, location=location)
 
 
+class FeatureExperimental(FeatureCheckBase):
+
+    """Checks for experimental features"""
+
+    feature_registry = {}
+
+    def __init__(self, feature_name: str, introduced_version: str, extra_message: str = '',
+                 stable_version: T.Optional[str] = None):
+        super().__init__(feature_name, introduced_version, extra_message)
+        self.stabilized = stable_version
+
+    @staticmethod
+    def get_warning_str_prefix(tv: str) -> str:
+        return 'Experimental features used:'
+
+    @staticmethod
+    def get_notice_str_prefix(tv: str) -> str:
+        return ''
+
+    @staticmethod
+    def check_version(target_version: str, feature_version: str) -> bool:
+        # For deprecation checks we need to return the inverse of FeatureNew checks
+        return not mesonlib.version_compare_condition_with_min(target_version, feature_version)
```

Comment:
The comment is, at minimum, copy-pasted and needs to specify what we actually want here. So: what do we want?

check_version() is supposed to return true whenever we don't need to emit a warning. If we do the same as a FeatureDeprecated, then that means given:
```meson
project('foo', meson_version: '>=1.0.0')
```
- an experimental feature introduced in 0.63.0 will not warn
- an experimental feature introduced in 1.0.0 or 1.2.0 will warn

If we assume a FeatureNew is added alongside every FeatureExperimental (which of course we should always do) then people using experimental features that are too new for their supported meson compatibility matrix will get:

- a warning that they are using features too new and which will break if people actually use the versions of meson they are permitted to use
- silence on the topic of this feature's stability


Another case is where users do not list a meson_version at all. In this case the Feature checks framework emits no information on the topic, since there is no target version. This means that users who do not use meson_version do not ever get told they are using unstable features.

Reply:
> Another case is where users do not list a meson_version at all. In this case the Feature checks framework emits no information on the topic, since there is no target version. This means that users who do not use meson_version do not ever get told they are using unstable features.

This is a longstanding bug, and it's out of scope for this PR to fix it. If a project doesn't set a project version they wont get *any* warnings (I guess FeatureBroken now? But even if so, that's relatively new). I've had https://github.com/mesonbuild/meson/pull/10520 opened for more than a year to solve this

Reply:
I guess we could wrap the FeatureNew check into the FeatureExperimental check itself.

Reply:
No, there's a distinctive difference and it's about user semantics.

The original design of the Feature checks framework was "if people don't say which versions they support, just DTRT and assume they are DTRT also". It's a valid argument to make that people could, in theory, not care about warnings. (I think that actually it makes more sense to have a semver slot for this. I took a quick stab at this at eli-schwartz/meson@7518f6895fce36104a5814ffaa749e450ae7966a.)

But stability warnings are something we always need to emit, or else we cannot actually claim they are unstable and we are not permitted to change its behavior in ways that break people's builds. It's not about letting people know they are using suboptimal stuff. It's about the fact that people have a *right* to be notified if something they are using can change its semantics or be deleted from meson in a point release. ;)

Reply:
> The original design of the Feature checks framework was "if people don't say which versions they support, just DTRT and assume they are DTRT also". It's a valid argument to make that people could, in theory, not care about warnings -- in fact, someone originally made that argument, apparently.

Yeah, it's an argument I don't buy. We should just set the default for meson_version to `>= '1.0.0'` and be done with it. Not fixing it also means that we have to constantly work around the fact that we're not setting a minimum, but then expect certain parts of the framework to warn anyway. Either not setting a version means "Don't warn me about things" or we should set it to a reasonable minimum, say the current major semver